source/slang/slang-ir-specialize-resources.cpp

```diff
@@ -171,6 +171,27 @@ struct ResourceParameterSpecializationContext
         return anySpecializableParam;
     }
 
+    // Returns true if `type` is an `IRStructType` with array-typed fields.
+    bool isStructTypeWithArray(IRType* type)
```

Comment:
Hypothetically, we could cache the results of this check on the `IRType` by introducing a decoration like `[containsArrays(true)]` or `[containsArrays(false)]`. That would avoid the need to re-run this recursive analysis on the same type over and over again.